File: android-native/app/src/main/java/com/example/kpkn/services/workout/ActiveWorkoutHolder.kt
package com.example.kpkn.services.workout

import com.example.kpkn.screens.workout.WorkoutViewModel
import java.lang.ref.WeakReference

object ActiveWorkoutHolder {
    @Volatile
    private var _viewModelRef: WeakReference<WorkoutViewModel>? = null

    fun set(viewModel: WorkoutViewModel) {
        _viewModelRef = WeakReference(viewModel)
    }

    fun clear() {
        _viewModelRef?.clear()
        _viewModelRef = null
    }

    fun get(): WorkoutViewModel? = _viewModelRef?.get()

    fun isActive(): Boolean = _viewModelRef?.get() != null

    fun handleAction(action: TimerAction) {
        _viewModelRef?.get()?.handleTimerAction(action)
    }
}

sealed class TimerAction {
    data object CompleteSet : TimerAction()
    data object SkipTimer : TimerAction()
    data object AddTime : TimerAction()
    data object SubtractTime : TimerAction()
}

File: ios-native/KPKNFit/KPKNFit/Services/Workout/ActiveWorkoutHolder.swift
import Foundation

final class ActiveWorkoutHolder {
    static let shared = ActiveWorkoutHolder()

    private weak var _viewModel: WorkoutViewModel?

    private init() {}

    func set(_ viewModel: WorkoutViewModel) {
        _viewModel = viewModel
    }

    func clear() {
        _viewModel = nil
    }

    func get() -> WorkoutViewModel? {
        _viewModel
    }

    func isActive() -> Bool {
        _viewModel != nil
    }

    func handleAction(_ action: TimerAction) {
        Task { @MainActor in
            _viewModel?.handleTimerAction(action)
        }
    }
}
